### PR TITLE
[EXPERIMENTAL] Autogenerated unittests - evaluating usability

### PR DESCRIPTION
**Pull request by JanKrivanek**
### Context
Generated with very early preview of unit tests generation tool. Trying to evaluate if there is any hope of that being useful and what would need to be worked on to achieve usefulness.

### Technical data

Baseline (main):
```
Test summary: total: 85, failed: 3, succeeded: 81, skipped: 1, duration: 6.9s
```

Target (generated):
```
Test summary: total: 690, failed: 56, succeeded: 634, skipped: 0, duration: 6.0s
```

| AssemblyName                     | LineRateBefore | LineRateAfter | LineRateDiff | BranchRateBefore | BranchRateAfter | BranchRateDiff |
| -------------------------------- | -------------- | ------------- | ------------ | ---------------- | --------------- | -------------- |
| StructuredLogger                 | 19.61 %        | 26.72 %       | 7.11 %       | 15.69 %          | 18.63 %         | 2.93 %         |
| StructuredLogger.Utils           | 0.00 %         | 9.47 %        | 9.47 %       | 0.00 %           | 3.75 %          | 3.75 %         |

### More comments
 * Still just heuristical version (so might not follow convention and references of the repo)
 * Very few and dummy code fixers - so we fallback a lot to commenting out tests with compilation diagnostic
 * No runtime feedback yet - so tests that crash/fail during runtime are kept

### CODE REVIEW

File: src/StructuredLogger.Tests/Analyzers/MessageTaskAnalyzerTests.cs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+using System.Linq;
+using Microsoft.Build.Logging.StructuredLogger;
+using Xunit;
+
+namespace Microsoft.Build.Logging.StructuredLogger.UnitTests
+{
+    /// <summary>
+    /// Minimal stub implementation of Task for unit testing.
+    /// </summary>
+    public class Task
```
Comment (JanKrivanek) [inline src/StructuredLogger.Tests/Analyzers/MessageTaskAnalyzerTests.cs:12]: This is useless, as it doesn't extend/inherit tested code. It actually yields the tests incompilable

File: src/StructuredLogger.Tests/Analyzers/ProjectReferenceGraphTests.cs [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+// //             // Arrange
+// //             var build = new DummyBuild();
+// //             var evalFolder = new DummyFolder("EvaluationFolder");
+// //             var projectEval = new DummyProjectEvaluation("C:/Project/Test.csproj");
+// //             // Do not add Items folder to project evaluation.
+// //             evalFolder.AddChild(projectEval);
+// //             build.EvaluationFolder = evalFolder;
+// // 
+// //             // Act & Assert
+// //             var exception = Record.Exception(() => new Microsoft.Build.Logging.StructuredLogger.ProjectReferenceGraph(build));
```
Comment (JanKrivanek) [inline src/StructuredLogger.Tests/Analyzers/ProjectReferenceGraphTests.cs:30]: The Record here conflicted between `Xunit` and `Microsoft.Build.Logging` namespaces (later got the precedence)
But this would be hard to catch

File: src/StructuredLogger.Tests/AssemblyLoadBuildEventArgs2Tests.cs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                "[Default]");
+
+            // Assert
+            Assert.Equal(expectedMessage, actualMessage);
+        }
+
+        /// <summary>
+        /// Tests that the parameterized constructor correctly assigns properties and that the Message property returns the formatted string when LoadingInitiator is provided.
+        /// </summary>
+        [Fact]
```
Comment (JanKrivanek) [inline src/StructuredLogger.Tests/AssemblyLoadBuildEventArgs2Tests.cs:81]: This and above should have been a data driven tests

File: src/StructuredLogger.Tests/BinaryLogTests.cs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+            string nullPath = null;
+            // Act & Assert
+            Assert.ThrowsAny<Exception>(() => BinaryLog.ReadRecords(nullPath));
+        }
+
+        /// <summary>
+        /// Tests that calling ReadRecords with a non-existent file path throws an exception.
+        /// </summary>
+        [Fact]
+        public void ReadRecords_String_NonExistentFile_ThrowsException()
```
Comment (JanKrivanek) [inline src/StructuredLogger.Tests/BinaryLogTests.cs:33]: This is a duplicate of `ReadBuild_String_NonExistentFile_ThrowsFileNotFoundException` with less specific exception - so it's superfluous

File: src/StructuredLogger.Tests/BinaryLogger/BinLogReaderTests.cs [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+using Microsoft.Build.Logging.StructuredLogger;
+using Moq;
+using Xunit;
+
+namespace Microsoft.Build.Logging.StructuredLogger.UnitTests
+{
+    /// <summary>
+    /// A fake Progress implementation to capture progress reports during tests.
+    /// </summary>
+    public class FakeProgress : Progress<ProgressUpdate>, Progress
```
Comment (JanKrivanek) [inline src/StructuredLogger.Tests/BinaryLogger/BinLogReaderTests.cs:17]: Again - 'hand-mocking' went wrong - it mixes System.Progress StructuredLogger.Progress

File: src/StructuredLogger.Tests/BinaryLogger/BuildEventArgsFieldsTests.cs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            // Assert
+            Assert.Equal(MessageImportance.Low, _buildEventArgsFields.Importance);
+        }
+
+        /// <summary>
+        /// Tests the getters and setters of all properties to ensure they store and retrieve values correctly.
+        /// This test sets all properties using sample values, including edge cases like empty and null strings.
+        /// </summary>
+        [Fact]
+        public void Properties_SetterAndGetter_ShouldWorkCorrectly()
```
Comment (JanKrivanek) [inline src/StructuredLogger.Tests/BinaryLogger/BuildEventArgsFieldsTests.cs:40]: This feels as a very low value test.
We might want to skip autoprops etc.

**Review by JanKrivanek — COMMENTED** [review]
quick pass to capture some obvious misses

**Comment by Janek91** [comment]
@JanKrivanek How did you generate these tests?

**Comment by fivemeyestore** [comment]
Interesting solution to this problem! I've been working on similar challenges recently.

**Comment by fivemserverhostings** [comment]
This is a well-structured codebase. I appreciate the clean separation of concerns.
